src/app/wallets/pay-on-chain.ts

```diff
@@ -0,0 +1,348 @@
+  if (user instanceof Error) return user
+
+  const { twoFA } = user
+
+  const twoFACheck = twoFA?.secret
+    ? await checkAndVerifyTwoFA({
+        walletId: senderWalletId,
+        amount: checkedAmount,
+        twoFASecret: twoFA.secret,
+        twoFAToken,
+      })
+    : true
+  if (twoFACheck instanceof Error) return twoFACheck
+
+  return payOnChainByWalletId({
+    senderWalletId,
+    amount,
+    address,
+    targetConfirmations,
+    memo,
+    sendAll,
+  })
+}
+
+export const payOnChainByWalletId = async ({
+  senderWalletId,
+  amount,
+  address,
+  targetConfirmations,
+  memo,
+  sendAll,
+}: PayOnChainByWalletIdArgs): Promise<PaymentSendStatus | ApplicationError> => {
+  const onchainLogger = baseLogger.child({
+    topic: "payment",
+    protocol: "onchain",
+    transactionType: "payment",
+    address,
+    amount,
+    memo,
+  })
+  const checkedAddress = checkedToOnChainAddress({
+    network: BTC_NETWORK,
+    value: address,
+  })
+  if (checkedAddress instanceof Error) return checkedAddress
+
+  const checkedTargetConfirmations = checkedToTargetConfs(targetConfirmations)
+  if (checkedTargetConfirmations instanceof Error) return checkedTargetConfirmations
+
+  const checkedAmount = sendAll
+    ? await getBalanceForWalletId(senderWalletId)
+    : checkedToSats(amount)
+  if (checkedAmount instanceof Error) return checkedAmount
+
+  const wallets = WalletsRepository()
+  const recipientWallet = await wallets.findByAddress(checkedAddress)
+  const isIntraLedger = !(recipientWallet instanceof Error)
+
+  if (isIntraLedger)
+    return executePaymentViaIntraledger({
+      senderWalletId,
+      recipientWalletId: recipientWallet.id,
+      amount: checkedAmount,
+      address: checkedAddress,
+      memo,
+      sendAll,
+      logger: onchainLogger,
+    })
+
+  return executePaymentViaOnChain({
+    senderWalletId,
+    amount: checkedAmount,
+    address: checkedAddress,
+    targetConfirmations: checkedTargetConfirmations,
+    memo,
+    sendAll,
+    logger: onchainLogger,
+  })
+}
+
+const executePaymentViaIntraledger = async ({
+  senderWalletId,
+  recipientWalletId,
+  amount,
+  address,
+  memo,
+  sendAll,
+  logger,
+}: {
+  senderWalletId: WalletId
+  recipientWalletId: WalletId
+  amount: Satoshis
+  address: OnChainAddress
+  memo: string | null
+  sendAll: boolean
+  logger: Logger
+}): Promise<PaymentSendStatus | ApplicationError> => {
+  if (recipientWalletId === senderWalletId) return new SelfPaymentError()
+
+  const intraledgerLimitCheck = await checkIntraledgerLimits({
+    amount,
+    walletId: senderWalletId,
+  })
+  if (intraledgerLimitCheck instanceof Error) return intraledgerLimitCheck
+
+  const usdPerSat = await getCurrentPrice()
+  if (usdPerSat instanceof Error) return usdPerSat
+
+  const fee = toSats(0)
+  const sats = toSats(amount + fee)
+  const usd = sats * usdPerSat
+  const usdFee = fee * usdPerSat
+
+  const getUserName = async (walletId: WalletId) => {
```

Comment:
so far we have been spelling username as `Username` (no capital on N)

Reply:
Fixed https://github.com/GaloyMoney/galoy/pull/911/commits/7a9d01b554c4b98e50409140ed6c3014f859763d